src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java

```diff
@@ -0,0 +1,187 @@
+import ru.practicum.shareit.user.exception.UserNotFoundException;
+
+import javax.transaction.Transactional;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class ItemServiceImpl implements ItemService {
+    private final UserRepository userRepository;
+    private final ItemRepository itemRepository;
+    private final BookingRepository bookingRepository;
+    private final CommentRepository commentRepository;
+
+    @Override
+    @Transactional
+    public ItemDto create(Long userId, ItemDto itemDto) throws UserNotFoundException {
+        User owner = userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException("user not found"));
+        Item item = ItemMapper.toItem(itemDto);
+        item.setOwner(owner);
+        item = itemRepository.save(item);
+        return ItemMapper.toItemDto(item);
+    }
+
+    @Override
+    @Transactional
+    public ItemDto update(Long userId, Long itemId, ItemDto itemDto)
+            throws ItemNotFoundException, UserNotFoundException {
+
+        User owner = userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException("user not found"));
+        Item repoItem = itemRepository.findById(itemId).orElseThrow(() -> new ItemNotFoundException("item not found"));
+        if (!repoItem.getOwner().getId().equals(owner.getId())) throw new ItemNotFoundException("item not found");
+
+        itemDto.setId(itemId);
+        Item item = ItemMapper.matchItem(itemDto, repoItem);
+        item.setOwner(owner);
+
+        itemRepository.save(item);
+        item = itemRepository.findById(itemId).orElseThrow(() -> new ItemNotFoundException("item not found"));
+
+        return ItemMapper.toItemDto(item);
+    }
+
+    @Override
+    public ItemDto get(Long userId, Long itemId) throws ItemNotFoundException, UserNotFoundException {
+        User user = userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException("user not found"));
+        Item repoItem = itemRepository.findById(itemId).orElseThrow(() -> new ItemNotFoundException("item not found"));
+        User owner = repoItem.getOwner();
+
+        ItemDto itemDto = ItemMapper.toItemDto(repoItem);
+        itemDto.setOwner(owner.getId());
+
+        List<Comment> commentList = commentRepository.findAllByItem_Id(itemId);
+        List<CommentDto> commentDtos = commentList.stream()
+                .map(CommentMapper::toCommentDto)
+                .collect(Collectors.toList());
+        itemDto.setComments(commentDtos);
+
+        if (!user.getId().equals(owner.getId())) return itemDto;
+
+        Sort sortDesc = Sort.by(Sort.Direction.DESC, "end");
+        Optional<Booking> lastBooking = bookingRepository.findTop1BookingByItem_IdAndEndIsBeforeAndStatusIs(
+                itemId, LocalDateTime.now(), Status.APPROVED, sortDesc);
+
+        itemDto.setLastBooking(lastBooking.isEmpty() ? null : LastBookingDto.builder()
```

Comment:
Можно заменить на выражение в функциональном стиле:
```suggestion
        itemDto.setLastBooking(lastBooking.map(booking -> LastBookingDto.builder()

Reply:
Пробовал, не работает